source/fundamentals/monitoring.txt

```diff
@@ -0,0 +1,409 @@
+``MongoClient`` object. The code then runs a find operation to test the
+``IsReadAndWrite`` class.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/Monitoring.java
+   :language: java
+   :dedent:
+   :start-after: start monitor-cluster-example
+   :end-before: end monitor-cluster-example
+
+The above code snippet should produce output that looks like this:
+
+.. code-block:: none
+   :copyable: false
+
+   Able to write to server
+   Able to read from server
+
+For more information on the classes and methods mentioned in this section, see
+the following API documentation:
+
+- :java-docs:`ClusterListener <apidocs/mongodb-driver-core/com/mongodb/event/ClusterListener.html>`
+- :java-docs:`ServerListener <apidocs/mongodb-driver-core/com/mongodb/event/ServerListener.html>`
+- :java-docs:`ServerMonitorListener <apidocs/mongodb-driver-core/com/mongodb/event/ServerMonitorListener.html>`
+- :java-docs:`MongoClientSettings <apidocs/mongodb-driver-core/com/mongodb/MongoClientSettings.html>`
+- :java-docs:`MongoClient <apidocs/mongodb-driver-sync/com/mongodb/client/MongoClient.html>`
+- :java-docs:`ClusterDescriptionChangedEvent <apidocs/mongodb-driver-core/com/mongodb/event/ClusterDescriptionChangedEvent.html>`
+
+Connection Pool Events
+~~~~~~~~~~~~~~~~~~~~~~
+
+A connection pool event is an event related to a **connection pool** held by the driver.
+A connection pool is a set of open TCP connections your driver maintains with
+a MongoDB instance. Connection pools help reduce the number of network handshakes
+your application needs to perform with a MongoDB instance, and can help your
+application run faster.
+
+To monitor connection pool events, write a class that implements the
+``ConnectionPoolListener`` interface and register an instance of that class with your
+``MongoClient`` instance.
+
+Example
+^^^^^^^
+
+This example shows how to make a listener class that prints a message each time
+you check out a connection from your connection pool.
+
+The following code defines the ``ConnectionPoolLibrarian`` class which implements the
+``ConnectionPoolListener`` interface.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/Monitoring.java
+   :language: java
+   :dedent:
+   :start-after: start cp-listener-impl
+   :end-before: end cp-listener-impl
+
+The following code adds an instance of the ``ConnectionPoolLibrarian`` class to a
+``MongoClient`` object. The code then runs a database command to test the
+librarian.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/Monitoring.java
+   :language: java
+   :dedent:
+   :start-after: start monitor-cp-example
+   :end-before: end monitor-cp-example
+
+The above code snippet should produce output that looks like this:
+
+.. code-block:: none
+   :copyable: false
+
+   Let me get you the connection with id 21...
+   Let me get you the connection with id 21...
+
+It looks like the above find operation requested two connections from the pool.
```

Comment:
That's not actually the reason.   Here's the stack trace for the second check-out:

```
connectionCheckedOut:159, ConnectionPoolLibrarian (com.mycompany.app)
get:123, DefaultConnectionPool (com.mongodb.internal.connection)
get:100, DefaultConnectionPool (com.mongodb.internal.connection)
getConnection:96, DefaultServer (com.mongodb.internal.connection)
endClosedSessions:156, ServerSessionPool (com.mongodb.internal.session)
close:103, ServerSessionPool (com.mongodb.internal.session)
close:110, MongoClientDelegate (com.mongodb.client.internal)
close:139, MongoClientImpl (com.mongodb.client.internal)
monitorConnectionPoolEvent:81, Monitoring (com.mycompany.app)
main:28, Monitoring (com.mycompany.app)
```

So it's happening because of a requirement that the driver informs the server about sessions that it has created and will no longer be using, since the client is being closed.

You can avoid this by not calling MongoClient.close()

Reply:
Thanks for catching this. Apologies for getting this incorrect. Here is the fix: https://github.com/mongodb/docs-java/pull/88/commits/c7ed0891dc61a14408a0085f0125838bc058ac9d

Going forward, I will try check these sorts of statements by consulting stack traces before including them in pages. Thanks for sharing this technique.

Reply:
It's ok.  You're learning!